### PR TITLE
Issue 6082: Upgrade to Bookkeeper 4.14

### PR DESCRIPTION
**Pull request by RaulGracia**
**Change log description**  
Upgrade Pravega to Bookkeeper 4.14.1.

**Purpose of the change**  
Fixes #6079.

**What the code does**  
This PR upgrades Bookkeeper to the latest stable version (4.14.1). Also, given that Bookkeeper 4.14 removes `bouncyCastle` dependency that was used by Pravega transitively, we included it explicitly in the dependencies. 

**How to verify it**  
- [x] Build passing.
- [x]  System tests (builds 3840, 3841).
- [x] Longevity run (2 days at 500MBps) worked fine (cc/ @OlegKashtanov).
- [x] Performance regression test (cc/ @AndreyKoltsov1997)

### CODE REVIEW

File: gradle.properties [hunk 1]
```diff
@@ -24,7 +24,7 @@ apacheCommonsCompressVersion=1.20
 apacheCuratorVersion=4.0.1
 apacheZookeeperVersion=3.5.5
 checkstyleToolVersion=8.23
-bookKeeperVersion=4.11.1
+bookKeeperVersion=4.14.0
```
Comment (eolivelli) [inline gradle.properties:27]: I suggest to go with 4.14.1

File: docker/bookkeeper/Dockerfile [hunk 2]
```diff
@@ -23,9 +23,9 @@
 # - github.com/apache/bookkeeper/blob/branch-4.9/docker/Dockerfile
 #
 
-FROM apache/bookkeeper:4.11.1
+FROM apache/bookkeeper:4.14.0
```
Comment (eolivelli) [inline docker/bookkeeper/Dockerfile:26]: 4.14.1

File: gradle.properties (hunk 1 above)
Comment (RaulGracia) [reply gradle.properties]: Done, thanks!

File: docker/bookkeeper/Dockerfile (hunk 2 above)
Comment (RaulGracia) [reply docker/bookkeeper/Dockerfile]: Done, thanks!

**Review by eolivelli — APPROVED** [review]
LGTM

File: build.gradle [hunk 3]
```diff
@@ -230,6 +230,9 @@ project('shared:security') {
         // in newer Java SE versions (9 and newer).
         compile group: 'javax.xml.bind', name: 'jaxb-api', version: jaxbVersion
         compile group: 'org.glassfish.jaxb', name: 'jaxb-runtime', version: jaxbVersion
+        // We need to explicitly import these as Bookkeeper 4.14+ is not transitively including them anymore.
+        compile group: 'org.bouncycastle', name: 'bcpkix-jdk15on', version: bouncyCastleVersion
```
Comment (andreipaduroiu) [inline build.gradle:234]: What do we use this for?

File: build.gradle [hunk 4]
```diff
@@ -786,6 +789,7 @@ project('cli:admin') {
         compile project(':segmentstore:server:host')
         compile project(':segmentstore:storage:impl')
         compile project(':shared:cluster')
+        compile project(":shared:security")
```
Comment (andreipaduroiu) [inline build.gradle:792]: Is this relevant to the upgrade?

File: build.gradle (hunk 3 above)
Comment (RaulGracia) [reply build.gradle:234]: In the previous version of Bookkeeper, we had these dependencies included transitively (`gradlew allDeps`):
```
    |         +--- org.bouncycastle:bcpkix-jdk15on:1.60
|    |         |    \--- org.bouncycastle:bcprov-jdk15on:1.60
|    |         +--- org.bouncycastle:bcprov-ext-jdk15on:1.60
```
However, in Bookkeeper 4.14, these are not included anymore:
```
|    |         +--- org.bouncycastle:bc-fips:1.0.2
```
The problem is that these are needed in Pravega with security enabled, as otherwise we get errors like when running tests or Pravega:
```
2021-06-10T09:48:16.2082894Z             Caused by:
2021-06-10T09:48:16.2083462Z             java.lang.NoSuchFieldError: xmss_with_SHA256
2021-06-10T09:48:16.2084627Z                 at org.bouncycastle.pqc.asn1.PQCObjectIdentifiers.<clinit>(Unknown Source)
2021-06-10T09:48:16.2127087Z                 at org.bouncycastle.jce.provider.BouncyCastleProvider.loadPQCKeys(Unknown Source)
2021-06-10T09:48:16.2129240Z                 at org.bouncycastle.jce.provider.BouncyCastleProvider.setup(Unknown Source)
```
Still, I will check in both are strictly necessary or we just need one of them.

File: build.gradle (hunk 4 above)
Comment (RaulGracia) [reply build.gradle:792]: Same problem as above.

**Comment by RaulGracia** [comment]
@eolivelli it looks like Bookkeeper `4.14.1` is not in DockerHub yet: https://hub.docker.com/r/apache/bookkeeper/tags?page=1&ordering=last_updated

**Comment by eolivelli** [comment]
@RaulGracia there should be a problem with the bot.
Can you please raise an issue in BookKeeper github ?
4.14.1 is not very different from 4.14.0 it just contains a fix for the Prometheus metrics (they were totally broken on 4.14.0)

**Comment by RaulGracia** [comment]
@eolivelli done: https://github.com/apache/bookkeeper/issues/2737

**Comment by eolivelli** [comment]
Thanks

**Comment by RaulGracia** [comment]
@andreipaduroiu we have done plenty of validations with this branch and we have not detected any problem with the upgrade.

**Comment by medvedevigorek** [comment]
@RaulGracia what about

**Review by andreipaduroiu — APPROVED** [review]
